FlowCrypt/Functionality/Services/Compose Message Service/ComposeMessageService.swift

```diff
@@ -0,0 +1,137 @@
+//
+//  ComposeMessageService.swift
+//  FlowCrypt
+//
+//  Created by Anton Kharchevskyi on 23.07.2021.
+//  Copyright © 2021 FlowCrypt Limited. All rights reserved.
+//
+
+import Combine
+import FlowCryptUI
+import Foundation
+
+struct ComposeMessageContext {
+    var message: String?
+    var recipients: [ComposeMessageRecipient] = []
+    var subject: String?
+}
+
+struct ComposeMessageRecipient {
+    let email: String
+    var state: RecipientState
+}
+
+protocol CoreComposeMessageType {
+    func composeEmail(msg: SendableMsg, fmt: MsgFmt, pubKeys: [String]?) throws -> CoreRes.ComposeEmail
+}
+
+final class ComposeMessageService {
+    private let messageGateway: MessageGateway
+    private let dataService: KeyStorageType
+    private let contactsService: ContactsServiceType
+    private let core: CoreComposeMessageType
+
+    init(
+        messageGateway: MessageGateway = MailProvider.shared.messageSender,
+        dataService: KeyStorageType = KeyDataStorage(),
+        contactsService: ContactsServiceType = ContactsService(),
+        core: CoreComposeMessageType = Core.shared
+    ) {
+        self.messageGateway = messageGateway
+        self.dataService = dataService
+        self.contactsService = contactsService
+        self.core = core
+    }
+
+    // MARK: - Validation
+    func validateMessage(
+        input: ComposeMessageInput,
+        contextToSend: ComposeMessageContext,
+        email: String,
+        atts: [SendableMsg.Attachment]
+    ) -> Result<SendableMsg, ComposeMessageError> {
+        let recipients = contextToSend.recipients
+
+        guard recipients.isNotEmpty else {
+            return .failure(.validationError(.emptyRecipient))
+        }
+
+        let emails = recipients.map(\.email)
+        let hasContent = emails.filter { $0.hasContent }
+
+        guard emails.isNotEmpty else {
+            return .failure(.validationError(.emptyRecipient))
+        }
+
+        guard emails.count == hasContent.count else {
+            return .failure(.validationError(.emptyRecipient))
+        }
+
+        guard input.isReply || contextToSend.subject?.hasContent ?? false else {
+            return .failure(.validationError(.emptySubject))
+        }
+
+        guard let text = contextToSend.message, text.hasContent else {
+            return .failure(.validationError(.emptyMessage))
+        }
+
+        let subject = input.subjectReplyTitle
+            ?? contextToSend.subject
+            ?? "(no subject)"
+
+        guard let myPubKey = self.dataService.publicKey() else {
+            return .failure(.validationError(.missedPublicKey))
+        }
+
+        switch getPubKeys(for: recipients) {
```

Comment:
Isn't there some mechanism like

```ts
let recipientPublicKeys = await getPubKeys(for: recipients)

let replyToMimeMsg = input.replyToMime
   .flatMap { String(data: $0, encoding: .utf8) }

let msg = SendableMsg(
    text: text,
    to: recipients.map(\.email),
    cc: [],
    bcc: [],
    from: email,
    subject: subject,
    replyToMimeMsg: replyToMimeMsg,
    atts: atts,
    pubKeys: allRecipientPubs + [myPubKey]
)
return .success(msg)
```

That would re-propagate the error from `getPubKeys` automatically?

Reply:
`getPubKeys` returns `Result<[String], MessageValidationError>`, it can be success or failure cases with some associated values in it. 
I can't use guard or if because I will miss one of 2 associated values
```
guard case .success(let allRecipientPubs) = getPubKeys(for: recipients) else {
           return ...(here I don't have associated error if getPubKeys fails)
}
```

It's possible to use `do catch` here, but I wouldn't recommend this solution, because we need to cast to some particular error in catch 
```
do {
            try getPubKeys(for: recipients).get()
        } catch let error {
            (here it's Error type, not MessageValidationError)
            print(error)
        }
```

best way is to use `Result<SendableMsg, ComposeMessageError>.Publisher` and then in view controller we can chain `validateMessage` with `encryptAndSend(message: sendableMessage)` functions. 
Please check latest commit.